### PR TITLE
QA differentiator test

### PR DESCRIPTION
**Pull request by DonaldJennings**
## Description
Contains the Tests for the Differentiator in utils/math. Tests the properties of differentiation, special cases, and some edge cases
 *Click up link* [](https://app.clickup.com/t/9gy84d)

## Changes
- Created differentiator.test.cpp
- Added Tests for Differentiator.hpp
- Added comments to differentiator.test.cpp explaining the tests.

## Additional Info 
Tests will be run on compile. All tests pass

### CODE REVIEW

File: test/src/utils/differentiator.test.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+{
+  ASSERT_EQ(diff_test.update(test_point).value, 0)
+    << "Initalised value should be 0";
+}
+
+/**
+ * Test fixture used for determining whether initialisation works correctly.
+ * The timestamp should always be returned unchanged
+*/
+TEST_F(DifferentiatorFunctionality, differentiatorInitialisedTimestamp)
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:98]: As a general comment for future tests, the name of the test should start with a verb. for example. handlesInitialTimestep

File: test/src/utils/differentiator.test.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+      << "The derivative should be the same as the difference of the two derivatives";
+  }
+}
+
+/**
+ * Struct used for testing that the chain rule applies in the
+ * differentiator. Creates 2 lists of DataPoints for the outer
+ * and inner functions and stores their DataPoints.
+ */
+struct DifferentiatorChainRule : public ::testing::Test {
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:297]: If the struct is only used once maybe then you can just use TEST instead of TEST_F

File: test/src/utils/differentiator.test.cpp [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+ */
+struct SpecialCases : public ::testing::Test {
+ protected:
+  Differentiator<float> diff_linear;
+  DataPoint<float> linear_data[100];
+  DataPoint<float> linear_data_point;
+  Differentiator<float> diff_constant;
+  DataPoint<float> function_constant[100];
+  DataPoint<float> constant_data_point;
+  const float kConstant = 10;
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp]: Maybe instead of a constant, we could try to get a random number here

File: test/src/utils/differentiator.test.cpp (hunk 2 above)
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp]: If the struct is only used once maybe then you can just use TEST instead of TEST_F

File: test/src/utils/differentiator.test.cpp (hunk 1 above)
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:98]: As a general comment for future tests, the name of the test should start with a verb. for example. handlesInitialTimestep

File: test/src/utils/differentiator.test.cpp [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+    second_point.timestamp = 2;
+
+    third_point.value= second_point.value;
+    third_point.timestamp = 3;
+  }
+
+  /**
+   * @brief Returns the difference between two data point values.
+  */
+  float deltaValue(DataPoint<float> point_1, DataPoint<float> prev_point)
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp]: Usually, Structs do not contain function/methods so maybe you can leave them outside as helper functions

File: test/src/utils/differentiator.test.cpp [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+/**
+ * Test fixture used for ensuring timestamp remains unchanged.
+ * Timestamp should always return unchanged whenever update is called.
+ */
+TEST_F(DifferentiatorFunctionality, differentiatorTimestampReturn)
+{
+  diff_test.update(test_point);
+  ASSERT_EQ(diff_test.update(second_point).timestamp, second_point.timestamp)
+    << "Timestamp should remain unchanged on Update";
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:123]: We should find a way of doing this better, maybe first encapsulate the error message in a variable, also maybe we can create a helper function for tests so that we can do the assertion and also pass the message as a parameter. Maybe changing the colour of the print would also be a good idea (red maybe)

File: test/src/utils/differentiator.test.cpp [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+ * Test For the Constant Case, we expect a perfect fit. As the gradient of this function should
+ * always be 0.
+ */
+TEST_F(SpecialCases, differentiatorConstantCase)
+{
+  diff_constant.update(function_constant[0]);
+  for (int i = 1; i < 100; i++) {
+    float value = diff_constant.update(function_constant[i]).value;
+    ASSERT_EQ(0, value)
+      << "You Expect a perfect fit for linear case, please review implementation";
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp]: This is not the linear case! this is the constant case (look the error message)

**Review by pmorande27 — APPROVED** [review]
In general, all is ok. Some small changes for future tests though.

File: test/src/utils/differentiator.test.cpp (hunk 4 above)
Comment (brano2) [reply test/src/utils/differentiator.test.cpp]: What if you have a case where you cannot move a method outside (e.g. it uses the instance variables)? Normally, you switch from a `struct` to a `class` in these cases. :sweat_smile:

Comment (DonaldJennings) [reply test/src/utils/differentiator.test.cpp]: Perhaps Classes could be more suited for building our Test Suites then since we can contain helper functions as methods within the class rather than making them all just global helper functions :)

Comment (pmorande27) [reply test/src/utils/differentiator.test.cpp]: mmmm okay I agree that classes are better, however, I am still convinced that you could do it do everything with enums and extract the functions. If you need to use the instance variables you could pass that as an argument and call the function in set Up or even in the test (as it has also access to them). If you want to modify them you could pass it as a pointer. But yeah that is only me being headstrong.

File: test/src/utils/differentiator.test.cpp [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+    ASSERT_EQ(function, quadratic-linear)
+      << "The derivative should be the same as the difference of the two derivatives";
+  }
+}
+
+/**
+ * Test fixture for checking the Chain Rule property of derivatives.
+ */
+TEST(DifferentiatorChainRule, differentiatorChainRule)
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:268]: This is ok now, other option would have been to merge the two structs into one

File: test/src/utils/differentiator.test.cpp [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+/**
+ * Struct used in Testing the difference of derivatives. Similar
+ * to the struct used for finding the sum except it finds the
+ * difference .
+ * Populates arrays of DataPoints with sample values drawn from the function
+ * outputs.
+ */
+struct DifferentiatorDifference : public ::testing::Test {
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp]: mmm I would say that you do not need this struct it is only used once

Comment (pmorande27) [reply test/src/utils/differentiator.test.cpp]: maybe this time merge Differentiatior difference with DiffProperty

File: test/src/utils/differentiator.test.cpp [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+ */
+struct SpecialCases : public ::testing::Test {
+ protected:
+  Differentiator<float> diff_linear;
+  DataPoint<float> linear_data[100];
+  DataPoint<float> linear_data_point;
+  Differentiator<float> diff_constant;
+  DataPoint<float> function_constant[100];
+  DataPoint<float> constant_data_point;
+  const float kConstant = rand() % 100;
```
Comment (pmorande27) [inline test/src/utils/differentiator.test.cpp:156]: This could be an integer I think as you are doing %

**Review by pmorande27 — APPROVED** [review]

**Review by flori-margaritescu — APPROVED** [review]

**Review by pmorande27 — APPROVED** [review]
All ok now, ready to merge
